### Exercise 4: TF-IDF (Term Frequency–Inverse Document Frequency) 

### Aim : 
To implement the TF-IDF (Term Frequency–Inverse Document Frequency) technique for converting textual data into weighted numerical features and identifying the importance of words in a document collection. 

### Objective:

1.Create a vocabulary from the given text corpus.

2.Calculate Term Frequency (TF) for each word in each document.

3.Calculate Document Frequency (DF) for each word.

4.Calculate Inverse Document Frequency (IDF).

5.Generate the TF-IDF matrix using Python.

6.Identify important words based on their TF-IDF scores.

### Expected Output:
A vocabulary containing all unique words.

TF values showing word frequency in each document.

DF and IDF values for each word.

A TF-IDF matrix containing numerical weights for each document and word.

Words with higher TF-IDF scores will be identified as more important or distinctive in a document.

### Pre-Lab 
Consider the following documents: 

D1: Apple is sweet

D2: Apple is red 

D3: Mango is sweet 

### 1.List all the unique words (Vocabulary)?
A. The unique words (vocabulary) are: I, love, NLP, Python, is, fun.
So, the vocabulary contains 6 unique words.

### 2.What is the size of the vocabulary?
A.The size of the vocabulary is 6.
It contains 6 unique words: I, love, NLP, Python, is, fun.

### 3.Calculate the Term Frequency (TF) and  write formula ?
A. Term Frequency (TF) is the number of times a word appears in a document divided by the total number of words in that document.
Formula: TF = (Number of times the term appears) / (Total number of terms in the document).

### 4. Write formula and Calculate the Document Frequency (DF) ?
A.Document Frequency (DF) is the number of documents in which a particular word appears at least once.
Formula: DF(term) = Number of documents containing the term.

### 5.Write formula and fine Inverse Document Frequency (IDF)? 
A.Inverse Document Frequency (IDF) measures how important or rare a word is across all documents.
Formula: IDF = log(Total number of documents / Number of documents containing the term).

### 6.Write the formula for calculating TF-IDF and give the table?
A.Formula: TF-IDF = TF × IDF.
Table: | Word | TF | IDF | TF-IDF |.

### In-Lab 

### Activity 1: Vocabulary Creation 

Tasks 
1. Import the required Python libraries.
2. Create the input corpus.
3. Generate the vocabulary.
4. Display the vocabulary.  

In [1]:
# Import library
import math
import pandas as pd

# Create input corpus
documents = [
    "I love NLP",
    "I love Python",
    "NLP is fun"
]

# Generate vocabulary
vocabulary = set()

for doc in documents:
    words = doc.lower().split()
    vocabulary.update(words)

vocabulary = sorted(vocabulary)

# Display vocabulary
print("Vocabulary:")
print(vocabulary)
print("Vocabulary Size:", len(vocabulary))

Vocabulary:
['fun', 'i', 'is', 'love', 'nlp', 'python']
Vocabulary Size: 6


### Activity 2: Term Frequency (TF), Document Frequency (DF) and Inverse Document Frequency (IDF), TF-IDF Matrix 
Tasks :
1. Calculate the total number of words in each document.  
2. Compute the Term Frequency (TF) of each word.  
3. Display the TF matrix. 
4. Calculate the Document Frequency (DF) of each term.  
5. Compute the Inverse Document Frequency (IDF). 
6. Display the DF and IDF values in tabular format. Display the TF-IDF matrix.  


In [3]:
# Import libraries
import math
import pandas as pd

# Input corpus
documents = [
    "I love NLP",
    "I love Python",
    "NLP is fun"
]

# Create vocabulary
vocabulary = sorted(set(word for doc in documents
                        for word in doc.lower().split()))

# Calculate total words in each document
total_words = [len(doc.lower().split()) for doc in documents]

# Calculate TF matrix
tf_matrix = []

for doc in documents:
    words = doc.lower().split()
    tf_row = []

    for word in vocabulary:
        tf = words.count(word) / len(words)
        tf_row.append(tf)

    tf_matrix.append(tf_row)

tf_df = pd.DataFrame(tf_matrix, columns=vocabulary,
                     index=["D1", "D2", "D3"])

print("Total words in each document:")
print(total_words)

print("\nTF Matrix:")
print(tf_df)

# Calculate DF
df_values = []

for word in vocabulary:
    df = sum(word in doc.lower().split() for doc in documents)
    df_values.append(df)

# Calculate IDF
N = len(documents)
idf_values = [math.log10(N / df) for df in df_values]

# Display DF and IDF
df_idf = pd.DataFrame({
    "Term": vocabulary,
    "DF": df_values,
    "IDF": idf_values
})

print("\nDF and IDF:")
print(df_idf)

# Calculate TF-IDF
tfidf_matrix = []

for i in range(len(documents)):
    row = []
    for j in range(len(vocabulary)):
        tfidf = tf_matrix[i][j] * idf_values[j]
        row.append(tfidf)
    tfidf_matrix.append(row)

tfidf_df = pd.DataFrame(tfidf_matrix,
                        columns=vocabulary,
                        index=["D1", "D2", "D3"])

print("\nTF-IDF Matrix:")
print(tfidf_df)

Total words in each document:
[3, 3, 3]

TF Matrix:
         fun         i        is      love       nlp    python
D1  0.000000  0.333333  0.000000  0.333333  0.333333  0.000000
D2  0.000000  0.333333  0.000000  0.333333  0.000000  0.333333
D3  0.333333  0.000000  0.333333  0.000000  0.333333  0.000000

DF and IDF:
     Term  DF       IDF
0     fun   1  0.477121
1       i   2  0.176091
2      is   1  0.477121
3    love   2  0.176091
4     nlp   2  0.176091
5  python   1  0.477121

TF-IDF Matrix:
        fun         i       is      love       nlp   python
D1  0.00000  0.058697  0.00000  0.058697  0.058697  0.00000
D2  0.00000  0.058697  0.00000  0.058697  0.000000  0.15904
D3  0.15904  0.000000  0.15904  0.000000  0.058697  0.00000


### Post Lab: Use the document and prepare Tf-IDF matrix

### D1: Python is easy 
### D2: Python supports machine learning 
### D3: Machine learning is powerful 

### Post Lab: TF-IDF Matrix

Given documents:

D1: Python is easy
D2: Python supports machine learning
D3: Machine learning is powerful
1. Vocabulary

[easy, is, learning, machine, powerful, python, supports]

Vocabulary size = 7

2. Total words in each document
Document	Total Words
D1	3
D2	4
D3	4
3. TF Matrix

Formula:

TF = Term frequency / Total words in document

Document	easy	is	learning	machine	powerful	python	supports
D1	0.333	0.333	0	0	0	0.333	0
D2	0	0	0.25	0.25	0	0.25	0.25
D3	0	0.25	0.25	0.25	0.25	0	0
4. DF and IDF

Formula:

IDF = log₁₀(N / DF)

Here, N = 3 documents.

Term	DF	IDF
easy	1	0.4771
is	2	0.1761
learning	2	0.1761
machine	2	0.1761
powerful	1	0.4771
python	2	0.1761
supports	1	0.4771
5. TF-IDF Matrix

TF-IDF = TF × IDF

Document	easy	is	learning	machine	powerful	python	supports
D1	0.1590	0.0587	0	0	0	0.0587	0
D2	0	0	0.0440	0.0440	0	0.0440	0.1193
D3	0	0.0440	0.0440	0.0440	0.1193	0	0